### PR TITLE
[kube-prometheus-stack] Add prometheus.thanosService.omitClusterIP

### LINKED ISSUE #863: [kube-prometheus-stack] prometheus.thanosService.clusterIP should not be required
**Describe the bug**

`prometheus.thanosService.clusterIP` ought not be a required value

**Version of Helm and Kubernetes**:

Helm Version:

```console
$ helm version --short
v3.5.3+g041ce5a
```

Kubernetes Version:

```console
$ kubectl version --short
Client Version: v1.19.9
Server Version: v1.19.9
```

**Which chart**: kube-prometheus-stack

**Which version of the chart**: 14.9.0

**What happened**:

```
UPGRADE FAILED: cannot patch "prometheus-operator-thanos-discovery" with kind Service: Service "prometheus-operator-thanos-discovery" is invalid: spec.clusterIP: Invalid value: "": field is immutable
```

**What you expected to happen**: successful no-op upgrade

**How to reproduce it** (as minimally and precisely as possible):
**Changed values of values.yaml** (only put values which differ from the defaults):

values.yaml

```console
prometheus:
  thanosService:
    enabled: true
    clusterIP: null # or ""
    type: LoadBalancer
```

**The helm command that you execute and failing/misfunctioning**:

For example:

```console
helm install prometheus-operator prometheus-community/kube-prometheus-stack -upgrade --version version --values values.yaml
```

**Anything else we need to know**:

This is essentially the same problem as https://github.com/helm/charts/issues/9227 (fixed in https://github.com/helm/charts/pull/13646)

### PR DESCRIPTION
**Pull request by yurrriq**
#### What this PR does / why we need it:

Support omission of the Thanos service `clusterIP`.

#### Which issue this PR fixes
  - fixes #863

#### Special notes for your reviewer:

Similar to https://github.com/helm/charts/pull/13646

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by yurrriq** [comment]
rebased. need anything else from me here? this is pretty much a blocker for us

**Comment by yurrriq** [comment]
rebased again.

File: charts/kube-prometheus-stack/values.yaml [hunk 1]
```diff
@@ -1597,6 +1597,7 @@ prometheus:
     portName: grpc
     port: 10901
     targetPort: "grpc"
+    omitClusterIP: false
```
Comment (monotek) [inline charts/kube-prometheus-stack/values.yaml:1600]: Why is this needed?
Imho you could just set `clusterIP: ""` to get rid of the error.

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/kube-prometheus-stack/values.yaml (hunk 1 above)
Comment (yurrriq) [reply charts/kube-prometheus-stack/values.yaml:1600]: I'll try that and report back.

Comment (yurrriq) [reply charts/kube-prometheus-stack/values.yaml:1600]: Same result.
```
UPGRADE FAILED: cannot patch "prometheus-operator-thanos-discovery" with kind Service: Service "prometheus-operator-thanos-discovery" is invalid: spec.clusterIP: Invalid value: "": field is immutable
```

File: charts/kube-prometheus-stack/templates/prometheus/serviceThanosSidecar.yaml [hunk 2]
```diff
@@ -16,7 +16,9 @@ metadata:
 {{- end }}
 spec:
   type: {{ .Values.prometheus.thanosService.type }}
+  {{- if not .Values.prometheus.thanosService.omitClusterIP }}
   clusterIP: {{ .Values.prometheus.thanosService.clusterIP }}
```
Comment (yurrriq) [inline charts/kube-prometheus-stack/templates/prometheus/serviceThanosSidecar.yaml:20]: As discussed in the linked tickets (see links in #863) regarding the nginx chart, the ideal solution here would be something like this, but if we want to maintain backward compatibility, we'll probably need something like `omitClusterIP`.

```suggestion
  {{- with .Values.prometheus.thanosService.clusterIP }}
  clusterIP: {{ . }}
  {{- end }}
```
```

**Comment by doug-ba** [comment]
I don't know that this is the correct solution for #863 .  This service is intended as a headless service for service discovery.  I've made an alternate PR in https://github.com/prometheus-community/helm-charts/pull/916 which creates a new service for external  access and does not set ClusterIP.

**Comment by yurrriq** [comment]
Superseded by #916
